File: sample-application-optional/src/main/java/de/roamingthings/workbench/optionalgreeterapp/OptionalGreeterApplication.java
package de.roamingthings.workbench.optionalgreeterapp;

import de.roamingthings.greeter.Greeter;
import de.roamingthings.optionalgreeter.OptionalGreeter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

import java.util.Optional;

@SpringBootApplication
public class OptionalGreeterApplication implements CommandLineRunner {
    private static Logger log = LoggerFactory
            .getLogger(OptionalGreeterApplication.class);

    private final Greeter greeter;
    private final OptionalGreeter optionalGreeter;

    @Autowired
    public OptionalGreeterApplication(Greeter greeter, OptionalGreeter optionalGreeter) {
        this.greeter = greeter;
        this.optionalGreeter = optionalGreeter;
    }

    public static void main(String[] args) {
        SpringApplication.run(OptionalGreeterApplication.class, args);
    }

    @Override
    public void run(String... args) {
        String name = retrieveNameFrom(args).orElse("Jon Doe");

        String greetingMessage = greeter.greet(name);
        String optionalGreeterMessage = optionalGreeter.wishNiceDay(name);

        log.info(greetingMessage);
        log.info(optionalGreeterMessage);
    }

    private Optional<String> retrieveNameFrom(String[] args) {
        String name = null;
        if (args.length > 0) {
            name = args[0];
        }
        return Optional.ofNullable(name);
    }
}

File: sample-application-optional/src/main/kotlin/de/roamingthings/workbench/optionalgreeterapp/OptionalGreeterApplication.kt
package de.roamingthings.workbench.optionalgreeterapp

import de.roamingthings.greeter.Greeter
import de.roamingthings.optionalgreeter.OptionalGreeter
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.CommandLineRunner
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Bean


@SpringBootApplication
class OptionalGreeterApplication {

    private val log = LoggerFactory.getLogger(OptionalGreeterApplication::class.java)

    @Bean
    fun init(
            greeter: Greeter,
            optionalGreeter: OptionalGreeter,
            applicationContext: ApplicationContext,
            applicationArguments: ApplicationArguments) = CommandLineRunner {

        val name = retrieveNameFrom(applicationArguments) ?: "Jon Doe"

        val greetingMessage = greeter.greet(name)
        val optionalGreeterMessage = optionalGreeter.wishNiceDay(name)

        log.info(greetingMessage)
        log.info(optionalGreeterMessage)
    }

    private fun retrieveNameFrom(applicationArguments: ApplicationArguments): String? =
            if (applicationArguments.sourceArgs.isNotEmpty()) applicationArguments.sourceArgs[0] else null

}

fun main(args: Array<String>) {
    runApplication<OptionalGreeterApplication>(*args)
}
